pkg/types/tuple.go

```diff
@@ -0,0 +1,139 @@
+package types
+
+import (
+	"fmt"
+
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+// Object represents an FGA object.
+//
+// See https://openfga.dev/docs/concepts#what-is-an-object
+type Object struct {
+	Type string
+	ID   string
+}
+
+var _ Subject = (*Object)(nil)
+
+// GetType returns the object's assigned type.
+func (o Object) GetType() string {
+	return o.Type
+}
+
+// String returns this object as a string formatted as 'type:id'.
+func (o Object) String() string {
+	return fmt.Sprintf("%s:%s", o.GetType(), o.ID)
+}
+
+func (o Object) isSubject() {}
+
+// Userset reprents a set of FGA users/subjects.
+//
+// See https://openfga.dev/docs/concepts#what-is-a-user (specifically for `userset`)
+type Userset struct {
+	Type     string
+	ID       string
+	Relation string
+}
+
+var _ Subject = (*Userset)(nil)
+
+// GetType returns the userset's assigned type.
+func (u Userset) GetType() string {
+	return u.Type
+}
+
+// String returns this userset as a string formatted as 'type:id#relation'.
+func (u Userset) String() string {
+	return fmt.Sprintf("%s:%s#%s", u.GetType(), u.ID, u.Relation)
+}
+
+func (u Userset) isSubject() {}
+
+// TypedWildcardSubject represents a publicly assignable subject that denotes any
+// user/subject of the specified type is included in the set of subjects denoted
+// by the typed wildcard.
+type TypedWildcardSubject struct {
+	Type string
+}
+
+var _ Subject = (*TypedWildcardSubject)(nil)
+
+// String returns this typed wildcard as a string formatted as 'type:*'.
+//
+// '*' is used to denote the wildcard representing a type restriction granting
+// public access.
+func (t TypedWildcardSubject) String() string {
+	return tuple.TypedPublicWildcard(t.Type)
+}
+
+// GetType returns the typed wildcard's assigned type.
+func (t TypedWildcardSubject) GetType() string {
+	return t.Type
+}
+
+func (t TypedWildcardSubject) isSubject() {}
+
+// Subject represents an FGA user/subject.
+//
+// A Subject can be any one of an [Object], [TypedWildcardSubject], or
+// [Userset].
+//
+// See https://openfga.dev/docs/concepts#what-is-a-user
+type Subject interface {
```

Comment:
How do we feel about using this internal definition as a good point to start this gradual rename in internal code from `user` to `subject` so that we put an end to the stuttering of the overloaded term "user"?

Happy to rename fields like `subject_type`, `subject_id`, and `subject_relation` to `user_type`, `user_id`, `user_relation` if we feel it's not best, but given this is an internal API contract I thought it might be nice to start evolving that direction.